File: Tools/SagaTools/src/Resolver.cpp
/// @file Resolver.cpp
/// @brief Implementation of name → executable-path resolution. UTF-8 throughout.

#include "SagaTools/Resolver.h"
#include "SagaTools/Encoding.h"

#include <filesystem>
#include <sstream>

namespace SagaTools
{

namespace
{

// ─── Path helpers (UTF-8 aware) ──────────────────────────────────────────────

bool FileExists(const std::string& utf8Path)
{
    std::error_code ec;
    const auto p = PathFromUtf8(utf8Path);
    return std::filesystem::exists(p, ec)
        && std::filesystem::is_regular_file(p, ec);
}

bool ContainsPathSeparator(const std::string& s)
{
    return s.find('/') != std::string::npos || s.find('\\') != std::string::npos;
}

#if defined(_WIN32)
constexpr char kPathSeparator = ';';
const std::vector<std::string> kExeSuffixes = {".exe", ".bat", ".cmd", ""};
#else
constexpr char kPathSeparator = ':';
const std::vector<std::string> kExeSuffixes = {""};
#endif

/// `${VAR}` substitution against the process environment. Unset vars expand
/// to the empty string. Lookup goes through `GetEnvUtf8` so values containing
/// non-ASCII characters survive on Windows (where `getenv` returns ANSI).
std::string ExpandEnv(const std::string& in)
{
    std::string out;
    out.reserve(in.size());
    for (std::size_t i = 0; i < in.size(); )
    {
        if (in[i] == '$' && i + 1 < in.size() && in[i + 1] == '{')
        {
            const std::size_t close = in.find('}', i + 2);
            if (close != std::string::npos)
            {
                const std::string var = in.substr(i + 2, close - (i + 2));
                out.append(GetEnvUtf8(var));
                i = close + 1;
                continue;
            }
        }
        out.push_back(in[i++]);
    }
    return out;
}

std::string SearchOnPath(const std::string& executable)
{
    const std::string env = GetEnvUtf8("PATH");
    if (env.empty()) return {};
    std::stringstream ss(env);
    std::string dir;
    while (std::getline(ss, dir, kPathSeparator))
    {
        if (dir.empty()) continue;
        for (const auto& suffix : kExeSuffixes)
        {
            const std::string candidate = dir + "/" + executable + suffix;
            if (FileExists(candidate))
                return PathToUtf8(std::filesystem::absolute(PathFromUtf8(candidate)));
        }
    }
    return {};
}

std::string AbsoluteFromRegistry(const std::string& baseDir, const std::string& path)
{
    namespace fs = std::filesystem;
    const fs::path p = PathFromUtf8(path);
    if (p.is_absolute()) return PathToUtf8(p);
    return PathToUtf8(fs::weakly_canonical(PathFromUtf8(baseDir) / p));
}

} // namespace

// ─── Resolution ───────────────────────────────────────────────────────────────

Resolution Resolver::Resolve(const Registry& registry, const std::string& name) noexcept
{
    Resolution out;

    const std::string* raw = registry.FindToolRaw(name);
    if (!raw) return out;

    out.expandedValue = ExpandEnv(*raw);
    if (out.expandedValue.empty()) return out;

    if (ContainsPathSeparator(out.expandedValue))
    {
        const std::string abs = AbsoluteFromRegistry(registry.BaseDir(), out.expandedValue);
        if (FileExists(abs))
        {
            out.available    = true;
            out.resolvedPath = abs;
            out.source       = "registry-path";
        }
        return out;
    }

    const std::string hit = SearchOnPath(out.expandedValue);
    if (!hit.empty())
    {
        out.available    = true;
        out.resolvedPath = hit;
        out.source       = "PATH";
    }
    return out;
}

} // namespace SagaTools

File: Tools/SagaTools/src/resolver.rs
/// @file resolver.rs
/// @brief Path resolution - mirrors C++ Resolver.cpp exactly.
///        Uses Rust native PathBuf (fixes bugs #3/#4).

use std::env;
use std::path::{Path, PathBuf};
use crate::registry::Registry;

#[derive(Debug, Clone)]
pub struct Resolution {
    pub available: bool,
    pub resolved_path: String,
    pub source: String,
    pub expanded_value: String,
}

pub struct Resolver;

impl Resolver {
    pub fn resolve(registry: &Registry, name: &str) -> Resolution {
        let mut out = Resolution {
            available: false,
            resolved_path: String::new(),
            source: String::new(),
            expanded_value: String::new(),
        };
        
        let raw = match registry.find_tool_raw(name) {
            Some(r) => r,
            None => return out,
        };
        
        out.expanded_value = Self::expand_env(raw);
        if out.expanded_value.is_empty() {
            return out;
        }
        
        if Self::contains_path_separator(&out.expanded_value) {
            let abs = Self::absolute_from_registry(registry.base_dir(), &out.expanded_value);
            if abs.exists() && abs.is_file() {
                out.available = true;
                out.resolved_path = abs.to_string_lossy().to_string();
                out.source = "registry-path".to_string();
            }
            return out;
        }
        
        if let Some(hit) = Self::search_on_path(&out.expanded_value) {
            out.available = true;
            out.resolved_path = hit.to_string_lossy().to_string();
            out.source = "PATH".to_string();
        }
        
        out
    }
    
    fn expand_env(input: &str) -> String {
        let mut out = String::with_capacity(input.len());
        let bytes = input.as_bytes();
        let mut i = 0;
        
        while i < bytes.len() {
            if bytes[i] == b'$' && i + 1 < bytes.len() && bytes[i + 1] == b'{' {
                i += 2;
                let mut var_name = String::new();
                while i < bytes.len() && bytes[i] != b'}' {
                    var_name.push(bytes[i] as char);
                    i += 1;
                }
                if i < bytes.len() {
                    i += 1;
                }
                
                if let Ok(value) = env::var(&var_name) {
                    out.push_str(&value);
                } else if let Some(value) = env::var_os(&var_name) {
                    if let Some(value_str) = value.to_str() {
                        out.push_str(value_str);
                    }
                }
                continue;
            }
            out.push(bytes[i] as char);
            i += 1;
        }
        
        out
    }
    
    fn contains_path_separator(s: &str) -> bool {
        s.contains('/') || s.contains('\\')
    }
    
    fn absolute_from_registry(base_dir: &Path, path_str: &str) -> PathBuf {
        let p = Path::new(path_str);
        if p.is_absolute() {
            p.to_path_buf()
        } else {
            base_dir.join(p)
        }
    }
    
    fn search_on_path(executable: &str) -> Option<PathBuf> {
        let path_env = env::var("PATH").unwrap_or_default();
        
        #[cfg(windows)]
        let separator = ';';
        #[cfg(not(windows))]
        let separator = ':';
        
        #[cfg(windows)]
        let suffixes = vec![".exe", ".bat", ".cmd", ""];
        #[cfg(not(windows))]
        let suffixes: Vec<&str> = vec![""];
        
        for dir in path_env.split(separator) {
            if dir.is_empty() {
                continue;
            }
            for suffix in &suffixes {
                let candidate = Path::new(dir).join(format!("{}{}", executable, suffix));
                if candidate.exists() && candidate.is_file() {
                    return Some(candidate);
                }
            }
        }
        
        None
    }
}
